go/datas/database_common.go

```diff
@@ -96,34 +96,60 @@ func (ds *databaseCommon) doSetHead(datasetID string, commit types.Struct) error
 func (ds *databaseCommon) doCommit(datasetID string, commit types.Struct) error {
 	d.PanicIfTrue(!IsCommitType(commit.Type()), "Can't commit a non-Commit struct to dataset %s", datasetID)
 
-	currentRootRef, currentDatasets := ds.getRootAndDatasets()
-	commitRef := ds.WriteValue(commit) // will be orphaned if the tryUpdateRoot() below fails
-
-	// First commit in store is always fast-forward.
-	if !currentRootRef.IsEmpty() {
-		r, hasHead := currentDatasets.MaybeGet(types.String(datasetID))
-
-		// First commit in dataset is always fast-forward.
-		if hasHead {
-			currentHeadRef := r.(types.Ref)
-			// Allow only fast-forward commits.
-			if commitRef.Equals(currentHeadRef) {
-				return nil
-			}
-			if !CommitDescendsFrom(commit, currentHeadRef, ds) {
-				return ErrMergeNeeded
+	var err error
+	for err = ErrOptimisticLockFailed; err == ErrOptimisticLockFailed; {
```

Comment:
can we put a comment here to a bug that we open that there's no guarentee about termination here? i.e. if a client gets unlucky and the database is under heavy load, it could theoretically stay in this loop forever. Eventually we probably need some kind of consensus mechanism which orders commits or something.